File: frontend/src/api/chartEditorStates.ts
import { apiClient } from "./client";

export interface ChartEditorStateRead {
  id: string;
  storage_key: string;
  form_id: string;
  chart_id: string;
  project_id: string;
  graphs_json: Record<string, unknown> | null;
  metadata_json: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface ChartEditorStateListRead {
  form_id: string;
  total: number;
  items: ChartEditorStateRead[];
}

export interface ChartEditorStateSave {
  storage_key: string;
  graphs_json: Record<string, unknown> | null;
  metadata_json?: Record<string, unknown> | null;
}

export function upsertChartEditorState(formId: string, chartId: string, payload: ChartEditorStateSave) {
  return apiClient.put<ChartEditorStateRead>(
    `/api/v1/forms/${formId}/chart-editor-states/${chartId}`,
    payload,
  );
}

export function listChartEditorStates(formId: string) {
  return apiClient.get<ChartEditorStateListRead>(`/api/v1/forms/${formId}/chart-editor-states`);
}

export function getChartEditorState(formId: string, chartId: string) {
  return apiClient.get<ChartEditorStateRead>(
    `/api/v1/forms/${formId}/chart-editor-states/${chartId}`,
  );
}

export function deleteChartEditorState(formId: string, chartId: string) {
  return apiClient.delete<{ status: string; chart_id: string }>(
    `/api/v1/forms/${formId}/chart-editor-states/${chartId}`,
  );
}

File: backend/app/routers/chart_editor_states.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.schemas.chart_editor_state import (
    ChartEditorStateDeleteResponse,
    ChartEditorStateListRead,
    ChartEditorStateRead,
    ChartEditorStateSave,
)
from app.services.chart_editor_state_service import ChartEditorStateService


router = APIRouter(tags=["chart-editor-states"])


def get_service(db: Session = Depends(get_db)) -> ChartEditorStateService:
    return ChartEditorStateService(db)


@router.put(
    "/api/v1/forms/{form_id}/chart-editor-states/{chart_id}",
    response_model=ChartEditorStateRead,
    status_code=200,
)
def upsert_chart_editor_state(
    form_id: str,
    chart_id: str,
    payload: ChartEditorStateSave,
    service: ChartEditorStateService = Depends(get_service),
) -> ChartEditorStateRead:
    return service.upsert(form_id, chart_id, payload)


@router.get(
    "/api/v1/forms/{form_id}/chart-editor-states",
    response_model=ChartEditorStateListRead,
)
def list_chart_editor_states(
    form_id: str,
    service: ChartEditorStateService = Depends(get_service),
) -> ChartEditorStateListRead:
    return service.list_for_form(form_id)


@router.get(
    "/api/v1/forms/{form_id}/chart-editor-states/{chart_id}",
    response_model=ChartEditorStateRead,
)
def get_chart_editor_state(
    form_id: str,
    chart_id: str,
    service: ChartEditorStateService = Depends(get_service),
) -> ChartEditorStateRead:
    return service.get(form_id, chart_id)


@router.delete(
    "/api/v1/forms/{form_id}/chart-editor-states/{chart_id}",
    response_model=ChartEditorStateDeleteResponse,
)
def delete_chart_editor_state(
    form_id: str,
    chart_id: str,
    service: ChartEditorStateService = Depends(get_service),
) -> ChartEditorStateDeleteResponse:
    return service.delete(form_id, chart_id)
